### ISSUE TITLE
Texture atlas it improvement

### ISSUE THREAD

**Issue by eqprog** [hrydgard/ppsspp#13133]
I’m new to github so I’m sorry if this is the wrong place to post this.
I’m not much of a programmer (especially with graphics) so I’m not sure how feasible this is. Please don’t interpret any of this post as negative!

I’m currently making a complete texture replacement mod for FF7: Crisis Core (on my github) but I’m not really here to plug that project. I’m using AI to upscale every texture in the game to 4x native resolution... currently I’m at 3GB of PNGs or just under 6000 files.

Performance really isn’t as bad as you might think, but I do have a good SSD and a 1080 with 8GB of ram. But some of my testers have had issues.

However, what IS inconvenient from my perspective as a creator is how the emulator dumps and replaces the textures - especially textures that are generated using an atlas (?)

You see, AI upscaling has trouble with working with alpha channels and there aren’t very many good models for the kind of images (both style and resolution) I’m working with. Point is, manual editing has to be done with many of the resulting textures.

![49230A75-813F-476B-9FD5-D4DA9C929364](https://user-images.githubusercontent.com/66541216/87255490-d6e54980-c450-11ea-85e3-b26b656482c4.png)
![63C2C71F-771B-4CCB-8093-EB08007C7EBD](https://user-images.githubusercontent.com/66541216/87255492-d77de000-c450-11ea-9971-297a8c30867c.png)

These textures were ripped from the game models. Just two textures. But I have like 30 different texture dumps for this model because of the atlasing/mapping. Meaning any edits I have to do, I now have to do 30x. There are about 15 or so models in the game where this is the case. Eventually, I figured out that I could set up an artboard in photoshop to make it easier to export all the changes at once, but it’s not very common knowledge. I’ve seen more than a few texture replacements that have no facial animations because for many people it’s just too much work.

I think the way this is being done could be better, basically. The emulator/game already knows how to take part of a texture and map it on to another. My understanding is that the emulator dumps the resulting image because that is what is in memory. I understand why it’s done this way, and it’s a pretty simple and generally effective solution.  But why can’t we have an option to upscale/replace the “base” textures only, inform the emulator the scaling of the replacement texture and have the emulator scale the coordinates being used for the atlasing and then combine the two?

Or, I know there is a way to crop a dumped texture to deal with the ^2 dimension requirement for psp games. Could it be modified to crop multiple areas of a texture? EG create splices of a texture and then only replace the section whose data actual changes?

Again, I have no idea how practical it is, but coming to a better solution would make things much easier for the end users actually creating the replacement textures.

**Comment by iota97** [comment]
I don't own that game myself so I'm not totally sure what you mean with 30 different texture, if it's about mipmaps (same texture with different size) you may try to set `ignoreMipmap = true` in texture.ini

About the alpha channel I wonder if you can just set up a script to apply the alpha channel from original texture (scaled with whatever filter) to the AI upscaled one?

For atlasing I have no idea how you may handle it myself beside with lots of patience I'm afraid.

**Comment by efonte** [comment]
These are some of the dumped textures:
![000000004bb54918ac84fc89](https://user-images.githubusercontent.com/3957463/87280265-3fd0ce00-c4f2-11ea-8839-1732227b98e4.png)
![000000004bb54918ad81c70f](https://user-images.githubusercontent.com/3957463/87280272-40696480-c4f2-11ea-9cf1-1abf4131f124.png)
![000000004bb54918b6d94e25](https://user-images.githubusercontent.com/3957463/87280273-4101fb00-c4f2-11ea-9fc9-45c64808e63f.png)
![000000004bb54918b314702a](https://user-images.githubusercontent.com/3957463/87280278-419a9180-c4f2-11ea-8d78-d3e07e30a5b7.png)

**Comment by efonte** [comment]
Metal Gear Solid: Peace Walker:
![00000000b491eb495295317d](https://user-images.githubusercontent.com/3957463/87289310-c4c0e500-c4fc-11ea-97b4-61c990673685.png)
![00000000de0bbd435295317d](https://user-images.githubusercontent.com/3957463/87289313-c5f21200-c4fc-11ea-91c1-f66f488bd482.png)
![0000000013c5adbb5295317d](https://user-images.githubusercontent.com/3957463/87289317-c68aa880-c4fc-11ea-97d8-ae227021e377.png)

**Comment by eqprog** [comment]
Thanks for the response. I can’t pretend to totally understand it, but I know enough about the subject to understand why it might not ever be practical. 

On the off chance another texture artist finds themselves reading this article, this is what I’m doing with photoshop to work around the issue. This is MUCH faster than creating an action and applying it through batch processing. 

First, take one of the “base” textures. Mask out the areas with animations, leaving only the static areas visible. Make sure to have a transparent background:

![4B41B1D7-D59A-4DAA-A445-890DD6EE0806](https://user-images.githubusercontent.com/66541216/87336324-49186580-c507-11ea-96fd-c289c202822a.png)

Save as a separate PSD document.

In the layer panel, right click and select “convert to smart object”, the “convert to linked” and save it as a “.psb” file. You should now be able to double click the thumbnail for your smart object and it will open in a separate tab/window. You can edit the smart object as you wish, and when you save it, it will update the contents in the main document.  (You can manually update it as well).

Go open up an explorer/finder window and navigate to wherever your dumped textures are. Select the textures you need and DRAG them into your project (not the smart object document!). You should press enter to place each file into the document. Notice how the layer names match the file names (this is key).

Now, make one copy of your smart object per texture you have and organize your layers Into pairs so that each texture file has the smart object above it. Make a group from each pair, copy each texture layer name and change paste it into its group name.  

For each group, right click and select “create artboard from group”. The artboard should have the groups name.

Now, if you export your file through the “Export As” dialogue you should see all that every artboard is being exported as its own image with your texture file’s name.

**Comment by unknownbrackets** [comment]
On the different facial expressions: you might be surprised.  There's this mentality that games are 100 GB downloads now just because they are so amazing and massive, but a lot of time it's because there's no reason for anyone to care about optimizing or keeping the size down.  Obviously it's also true that graphics have gotten exponentially larger too.

Even for PSP games, this happens.  For example, I know of multiple games that store exactly the same image separately, multiple times on the disc.  In their defense, there could be legitimate disc speed reasons to do this for UMD games - but not for the PSN versions.

Anyway, it's very likely (and unfortunate) these images are completely separate, even within the game's data.  Likely, the game just swaps its texture or texture pointer when the facial expression changes, so they duplicated all the other image content for convenience.

-[Unknown]

**Comment by Gamemulatorer** [comment]
@eqprog how about this now?

**Comment by hrydgard** [comment]
No, nothing has changed that will help this, except performance (and memory usage if you use the new formats). The atlasing problem remains.

**Comment by Jaffacakelover** [comment]
I wonder if this is causing me a similar issue? **Metal Gear Solid: Digital Graphic Novel** is dumping excessive amounts of textures with different hashes. Hundreds, sometimes _thousands_ of the same texture... although in this case, all copies of it look exactly the same.

![09ec8be06111acd6bafc27f5](https://github.com/hrydgard/ppsspp/assets/3490850/f1ca05c4-32fc-4609-b2b6-3d1232cf0137)

![2023-08-12 08_44_19-new](https://github.com/hrydgard/ppsspp/assets/3490850/cc902c9d-154c-48ca-998b-965830987284)
